modules/lsp.nix

```
@@ -0,0 +1,102 @@
+{
+  lib,
+  config,
+  ...
+}:
+let
+  inherit (lib) mkOption types;
+
+  cfg = config.lsp;
+
+  serverType = types.submodule {
+    options = {
+      enable = lib.mkEnableOption "the language server";
+
+      config = mkOption {
+        type = with types; attrsOf anything;
+        description = ''
+          Configurations for each language server.
+        '';
+        default = { };
+        example = {
+          cmd = [
+            "clangd"
+            "--background-index"
+          ];
+          root_markers = [
+            "compile_commands.json"
+            "compile_flags.txt"
+          ];
+          filetypes = [
+            "c"
+            "cpp"
+          ];
+        };
+      };
+    };
+  };
+in
+{
+  options.lsp = {
+    servers = mkOption {
+      type = types.attrsOf serverType;
+      description = ''
+        LSP servers to enable and/or configure.
+
+        This option is implemented using neovim's `vim.lsp` lua API.
+        If you prefer to use [nvim-lspconfig], see [`plugins.lsp`].
+
+        [nvim-lspconfig]: https://github.com/neovim/nvim-lspconfig
+        [`plugins.lsp`]: https://nix-community.github.io/nixvim/plugins/lsp/index.html
+      '';
+      default = { };
+      example = {
+        "*".config = {
+          root_markers = [ ".git" ];
+          capabilities.textDocument.semanticTokens = {
+            multilineTokenSupport = true;
+          };
+        };
+        luals.enable = true;
+        clangd = {
+          enable = true;
+          config = {
+            cmd = [
+              "clangd"
+              "--background-index"
+            ];
+            root_markers = [
+              "compile_commands.json"
+              "compile_flags.txt"
+            ];
+            filetypes = [
+              "c"
+              "cpp"
+            ];
+          };
+        };
+      };
+    };
+  };
+
+  config = {
+    extraConfigLua =
+      let
+        serversToEnable = lib.pipe cfg.servers [
+          (lib.filterAttrs (_: props: props.enable))
+          builtins.attrNames
+        ];
+      in
+      ''
+        do
+          local __nixvimLspServersConfigs = ${lib.nixvim.toLuaObject cfg.servers}
+
+          for server_name, server_props in ipairs(__nixvimLspServersConfigs) do
+            vim.lsp.config(server_name, server_props.config)
+          end
+
+          vim.lsp.enable(${lib.nixvim.toLuaObject serversToEnable})
+        end
+      '';
```

Since we're serialising the whole of `cfg.servers` anyway, we may as well move the `enable` into the loop too:

```suggestion
    extraConfigLua = ''
      do
        local __nixvimLspServers = ${lib.nixvim.toLuaObject cfg.servers}

        for server_name, server_props in ipairs(__nixvimLspServers) do
          vim.lsp.config(server_name, server_props.config)

          if server_props.enable then
            vim.lsp.enable(server_name)
          end
        end

      end
    '';
```